xapian-core/common/serialise-double.cc

```diff
@@ -90,13 +90,23 @@ string serialise_double(double v){
      * and reduce exp by 1, since this is the way doubles
      * are stored in IEEE-754.
      *
-     * conversion of mantissa to bits is done by multiplying
-     * mantissa with 2^26, extracting the integer part and
-     * or-ing it with result, then repeating this step once
-     * more.
+     * Conversion of mantissa to bits is done by
+     * multiplying the mantissa with 2^52, converting
+     * it to a 64 bit integer representation of the original
+     * double.
      */
+
+    static_assert((uint64_t)1 << 52 <= numeric_limits<double>::max(),
+		  "Check if 2^52 can be represented by a double");
```

Comment:
When you compare an integer to a `double`, first the integer gets converted to a `double`, then the two `double` values are compared (see "Conversions" section of https://en.cppreference.com/w/cpp/language/operator_arithmetic).  But we're trying to test if this integer is representable as a `double`, so we're relying on what happens if that conversion overflows.  If we'll definitely get an "overflow" or "infinity" value which doesn't compare <= the max value, all is well, but I'm struggling to work out if we can assume that's the case.

We could perhaps test based on the max exponent supported, but that's potentially complicated by what the radix is.  We currently hard require `FLT_RADIX` to be 2 in `xapian-core/api/sortable-serialise.cc`, but that requirement could be relaxed (2 seems to be ubiquitous for `double` on modern platforms, but 16 and 10 have been used on older machines).

So perhaps the best answer is to relax this check to `<` - in practice equality won't be achieved when we want the test to succeed, and by testing `<` we don't have to worry about that the conversion to `double` might give us `max()` as the overflow value if there's no infinity.

Reply:
Fixed accordingly.